### PR TITLE
Add UnionByName functionality

### LINKED ISSUE #284: Add unionByName visitor to update schema
### Feature Request / Improvement

This can be [ported from Java](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java).

### PR DESCRIPTION
**Pull request by Fokko**
Resolves #284

### CODE REVIEW

File: pyiceberg/table/__init__.py [hunk 1]
```diff
@@ -1398,14 +1401,23 @@ class UpdateSchema:
 
     def __init__(
         self,
-        table: Table,
+        table: Optional[Table],
```
Comment (HonahX) [inline pyiceberg/table/__init__.py:1404]: Is this change primarily for easier testing?

**Review by HonahX — COMMENTED** [review]
Thanks for the great work! @Fokko. It looks good. I left some comments and questions

File: tests/test_schema.py [hunk 2]
```diff
@@ -912,3 +914,668 @@ def test_promotion(file_type: IcebergType, read_type: IcebergType) -> None:
...
+
+def test_add_nested_lists(primitive_fields: NestedField) -> None:
+    new_schema = Schema(
+        NestedField(
+            field_id=1,
+            name="aList",
+            type=ListType(
+                element_id=2,
+                element_type=ListType(
+                    element_id=3,
```
[the remark this answers is not in the document]
Comment (Fokko) [reply tests/test_schema.py:1021]: It was the linter :D I guess it got the inspiration from Java: https://github.com/apache/iceberg/blob/17757276928e87e78314b6c19c8b3ebad9612619/core/src/test/java/org/apache/iceberg/TestSchemaUnionByFieldName.java#L429-L459 :D

File: pyiceberg/table/__init__.py [hunk 3]
```diff
@@ -1995,6 +2020,156 @@ def primitive(self, primitive: PrimitiveType) -> Optional[IcebergType]:
...
+
+    def schema(self, schema: Schema, partner_id: Optional[int], struct_result: bool) -> bool:
+        return struct_result
+
+    def struct(self, struct: StructType, partner_id: Optional[int], missing_positions: List[bool]) -> bool:
+        if partner_id is None:
+            return True
+
+        fields = struct.fields
+        partner_struct = self._find_field_type(partner_id)
```
[the remark this answers is not in the document]
Comment (Fokko) [reply pyiceberg/table/__init__.py:2041]: Good call, updated! 👍

File: pyiceberg/table/__init__.py [hunk 4]
```diff
@@ -1449,6 +1461,15 @@ def case_sensitive(self, case_sensitive: bool) -> UpdateSchema:
         self._case_sensitive = case_sensitive
         return self
 
+    def union_by_name(self, new_schema: Schema) -> UpdateSchema:
+        visit_with_partner(
+            new_schema,
+            -1,
+            UnionByNameVisitor(update_schema=self, new_schema=self._schema, case_sensitive=self._case_sensitive),  # type: ignore
```
[the remark this answers is not in the document]
Comment (Fokko) [reply pyiceberg/table/__init__.py:1468]: Ah good catch, I think that was an error, thanks for spotting this 👍

File: pyiceberg/table/__init__.py [hunk 5]
```diff
@@ -1995,6 +2020,156 @@ def primitive(self, primitive: PrimitiveType) -> Optional[IcebergType]:
...
+        if field.doc is not None and not field.doc != existing_field.doc:
+            self.update_schema.update_column(full_name, doc=field.doc)
+
+    def _find_field_type(self, field_id: int) -> IcebergType:
+        if field_id == -1:
+            return self.new_schema.as_struct()
+        else:
+            return self.new_schema.find_field(field_id).field_type
+
+    def field(self, field: NestedField, field_partner: Optional[int], field_result: bool) -> bool:
```
[the remark this answers is not in the document]
Comment (Fokko) [reply pyiceberg/table/__init__.py:2082]: I like it, thanks for the suggestion

File: pyiceberg/table/__init__.py [hunk 6]
```diff
@@ -1995,6 +2020,156 @@ def primitive(self, primitive: PrimitiveType) -> Optional[IcebergType]:
...
+                struct = self.partner_schema.find_field(partner_field_id).field_type
+                if not struct.is_struct:
+                    raise ValueError(f"Expected StructType: {struct}")
+
+            if field := struct.field_by_name(name=field_name, case_sensitive=self.case_sensitive):
+                return field.field_id
+
+        return None
+
+    def list_element_partner(self, partner_list: Optional[int]) -> Optional[int]:
```
[the remark this answers is not in the document]
Comment (Fokko) [reply pyiceberg/table/__init__.py:2148]: I like it 👍

File: pyiceberg/table/__init__.py (hunk 1 above)
Comment (Fokko) [reply pyiceberg/table/__init__.py:1404]: Yes, this was mostly for testing, but you can also use it without a table if you would

**Review by HonahX — APPROVED** [review]
Thanks @Fokko. LGTM!

File: pyiceberg/table/__init__.py [hunk 7]
```diff
@@ -1995,6 +2020,159 @@ def primitive(self, primitive: PrimitiveType) -> Optional[IcebergType]:
...
+        partner_map_type = self._find_field_type(map_partner_id)
+        if not isinstance(partner_map_type, MapType):
+            raise ValueError(f"Expected map-type, got: {partner_map_type}")
+
+        self._update_column(map_type.key_field, partner_map_type.key_field)
+        self._update_column(map_type.value_field, partner_map_type.value_field)
+
+        return False
+
+    def primitive(self, primitive: PrimitiveType, primitive_partner: Optional[int]) -> bool:
```
Comment (HonahX) [inline pyiceberg/table/__init__.py:2122]:
```suggestion
    def primitive(self, primitive: PrimitiveType, primitive_partner_id: Optional[int]) -> bool:
```
Just for consistency :)

Comment (Fokko) [reply pyiceberg/table/__init__.py]: Missed that one, thanks :)

**Comment by Fokko** [comment]
Thanks for the review @HonahX

**Pull request merged**